modules/images/webp-uploads/helper.php

```diff
@@ -144,7 +144,10 @@ function webp_uploads_generate_additional_image_source( $attachment_id, $image_s
 		$destination_file_name = $editor->generate_filename( $suffix, null, $extension[0] );
 	}
 
+	remove_filter( 'image_editor_output_format', 'webp_uploads_filter_image_editor_output_format', 10, 3 );
 	$image = $editor->save( $destination_file_name, $mime );
+	add_filter( 'image_editor_output_format', 'webp_uploads_filter_image_editor_output_format', 10, 3 );
```

Comment:
Can you clarify why you add this change? Would be great to add a one line comment on why the filter needs to be temporarily removed here.

I assume it is because here we output a WebP already if needed, so we shouldn't inside of it have the mapping of WebP? Which problems did you get during testing? However, I don't understand how exactly this causes a problem. If it's just about test failures, maybe it's our tests that need changes, but when thinking through the logic here, I can't see how having this filter run here would be problematic. Let me clarify by example:
1. User uploads JPEG.
2. Our `image_editor_output_format` callback fires and changes the default output for the JPEG to WebP, since that's the only one in the array.
3. Then our custom logic in `webp_uploads_create_sources_property()` runs. It shouldn't really do anything because the only MIME type to output is WebP which WordPress should have done already, so I think there may be a flaw in our existing logic. But even if it then went through the array of MIME types, it would only generate WebP (again).
4. In that case it would call `webp_uploads_generate_additional_image_source()` passing `image/webp` for the `$mime_type`, so that would then be passed to the `WP_Image_Editor::save()` method.
5. That would trigger the `image_editor_output_format` filter again, however passing `image/webp` to it. Our own callback would then (unnecessarily, to be fair) look at WebP output mappings and see that `image/webp` is an output mapping there, so it wouldn't tweak the list, so we would still get a WebP out of that.

In other words, while it may be okay to remove the filter here temporarily, I am questioning that that is really _needed_. We may have problems somewhere else though because this logic here should never run in the first place if WordPress only generates one MIME type.

I wonder if the problem is around https://github.com/WordPress/performance/blob/1.4.0/modules/images/webp-uploads/load.php#L42-L67: We look at the attachment MIME type, which will remain `image/jpeg` if the uploaded image is a JPEG. However, even the "full" image (`_wp_attached_file`) may already be a WebP image, if it was downscaled (`-scaled` file name suffix), because WP core then already ran the `image_editor_output_format` filter. So I think our assumption there that `get_post_mime_type()` is the MIME type of `get_attached_file()` is incorrect. We still need to call `get_post_mime_type()`, but only to know what was originally uploaded - that file may however be completely unused, if WordPress downscaled the image (which it would already generate a WebP for instead of JPEG). Because of that, we then set an incorrect source for `image/jpeg` which is actually a WebP image, and that then results in the `foreach` loop below to try to generate `image/webp` (again), since it doesn't find it in the sources. Did that maybe lead to the problem you were running into?

Reply:
@felixarntz, the problem is much simpler but to explain it, I need to show you an example. Let's assume we have the following image mime transformations:

```php
array(
    'image/jpeg' => array( 'image/webp' ),
    'image/webp' => array( 'image/webp', 'image/jpeg' ),
)
```

Such transformation leads to the following results:

```
    ["medium"]=>
    array(5) {
      ["file"]=>
      string(25) "balloons-175-200x300.webp"
      ["width"]=>
      int(200)
      ["height"]=>
      int(300)
      ["mime-type"]=>
      string(10) "image/webp"
      ["sources"]=>
      array(2) {
        ["image/webp"]=>
        array(2) {
          ["file"]=>
          string(25) "balloons-175-200x300.webp"
          ["filesize"]=>
          int(21246)
        }
        ["image/jpeg"]=>
        array(2) {
          ["file"]=>
          string(29) "balloons-175-200x300-webp.webp"
          ["filesize"]=>
          int(20417)
        }
      }
    }
```

Pay attention to the `image/jpeg` source, it is generated as webp image again even though we wanted to create a jpeg version. Why does it happen? Because our logic in the `webp_uploads_filter_image_editor_output_format` function knows nothing about which image is being generated and updates mime type only based on the transformation settings.

In other words, the `webp_uploads_filter_image_editor_output_format` function sees that the incoming image has the `image/jpeg` mime type and tries to find the corresponding transformation in the list of transformations for the `image/jpeg` mime type. Of course, it can find only the `image/webp` mime type in this case because we have `'image/jpeg' => array( 'image/webp' )` transformation for jpeg images. An early return obviously won't help in this situation because we have a case where we have multiple mime types for an image.

Yes, you can say that it won't be the case anymore because we removed `image/jpeg` from the allowed transformations list for webp images, but this can be incorrectly configured in a 3rd party plugin that use our filter to update transformations.

Reply:
@eugene-manuilov Yes, temporarily removing the filter makes sense. Still, I think we also need to have early returns here for the reason I outlined above: With the mapping of `image/jpeg` to `array( 'image/webp' )`, our plugin currently does way too much, and it tries to in fact generate another version of exactly the same WebP - so there are two problems here, the temporary filter removal addresses only one of them.

Let me explain, based on the current mapping:
* WordPress core, just via `webp_uploads_filter_image_editor_output_format()`, will generate WebP images for all the sizes of a JPEG upload, and even for the full / original image _if_ it gets downscaled.
* At this point, all necessary images are generated.
* We of course still need to populate `sources` for the attachment then, where our logic is already flawed: We call `get_post_mime_type()` which will be set to `image/jpeg`, but the full image _may_ in fact be `image/webp` (see above). We then store the existing full image source as `image/jpeg` although it is `image/webp`. This would then lead us to generate another "full" image WebP (basically the same that already exists), just because our logic thinks the existing image is a JPEG.

There may be a similar flaw in the sub-sizes, I haven't been able to assess that from code review yet, but we need to make sure we fix some of that logic so that we don't try to create the exact same WebP images again when they were already created by WordPress core.

Essentially: _If_ there is only one output MIME type, WordPress will have taken care of it. In that case, all our additional logic should do is populate the `sources` fields with those images, but not generate _any_ additional images. Does that make sense?

Reply:
Ok, yes, it does make sense. I have added one early return to avoid saving the same file to the disk:

https://github.com/WordPress/performance/pull/527/files#diff-1f653bd37f60a6476485a6794f7003a91b3713868c7414588d5835332fe68642R147-R153

Unfortunately, there is nothing else we can optimise in the code if only one mime type transformation is used.

Reply:
@eugene-manuilov Makes sense, after further inspection of the code I agree. With that said, there is still some unexpected behavior here. Will follow up in separate comments.